app/server/lib/GristJobs.ts

```diff
@@ -0,0 +1,318 @@
+import { makeId } from 'app/server/lib/idUtils';
+import log from 'app/server/lib/log';
+import { Queue, Worker } from 'bullmq';
+import IORedis from 'ioredis';
+
+/**
+ *
+ * Support for queues.
+ *
+ * We use BullMQ for queuing, since it seems currently the best the
+ * node ecosystem has to offer. BullMQ relies on Redis. Since queuing
+ * is so handy, but we'd like most of Grist to be usable without Redis,
+ * we make some effort to support queuing without BullMQ. This
+ * may not be sustainable, we'll see.
+ *
+ * Important: if you put a job in a queue, it can outlast your process.
+ * That has implications for testing and deployment, so be careful.
+ *
+ */
+export interface IGristJobs {
+  // All workers and jobs are scoped to individual named queues,
+  // with the real interfaces operating at that level.
+  queue(queueName?: string): IGristQueueScope;
+}
+
+/**
+ * For a given queue, we can add jobs, or methods to process jobs,
+ */
+export interface IGristQueueScope {
+  /**
+   * Add a job.
+   */
+  add(name: string, data: any, options?: JobAddOptions): Promise<void>;
```

Comment:
Is `data` needed? Can we start job just by its name?

Reply:
Yes, data in general is needed, it contains the parameters for the job. E.g. which document to make a thumbnail of, which batch of changes to accumulate in an email to which user, etc. Name and data are how BullMQ looks at this, other queue libraries see it as function name and arguments.